### PR TITLE
spl: Bump to newest SPL crates for v2.1 release

### PR DESCRIPTION
**Pull request by joncinque**
#### Problem

A lot of builds are currently broken due to the incompatibility between v2.0 and v2.1. All of the SPL crates have been bumped to use the Solana v2.1 crates, but it isn't being used in the monorepo.

#### Summary of changes

Bump the SPL crates to their newest versions, and correctly parse the new token-2022 instructions and account types.

The trickiest changes are the confidential transfer / mint / burn instruction parsing, otherwise the changes are straightforward.

Note: this needs to be backported to v2.1 so that users can use consistent crates. Currently, things are broken because of the major version bump of curve25519-dalek between v2.0 and v2.1. Once we break the circular dependency between the monorepo and SPL, this will be much easier.

### CODE REVIEW

**Comment by joncinque** [comment]
Downstream job failure will be fixed with https://github.com/solana-labs/solana-program-library/pull/7446

File: transaction-status/src/parse_token/extension/confidential_mint_burn.rs
Comment (samkim-crypto) [inline transaction-status/src/parse_token/extension/confidential_mint_burn.rs]: Should we explicitly check for the proof offset here and use `instructionSysvar`?

File: transaction-status/src/parse_token/extension/confidential_mint_burn.rs
Comment (samkim-crypto) [inline transaction-status/src/parse_token/extension/confidential_mint_burn.rs:59]: Here, we do not count the number of optional proof accounts. So should we use a safer `.get(i)` to index into the `account_indexes` for the proof instructions below?

File: transaction-status/src/parse_token/extension/confidential_mint_burn.rs
Comment (samkim-crypto) [inline transaction-status/src/parse_token/extension/confidential_mint_burn.rs]:
```suggestion
            if mint_data.equality_proof_instruction_offset != 0 {
                let account_index = *account_indexes.get(offset).ok_or(
                    ParseInstructionError::InstructionKeyMismatch(ParsableProgram::SplToken),
                )?;
                map.insert(
                    "equalityProofRecordAccount".to_string(),
                    json!(account_keys[account_index as usize].to_string()),
                );
                offset += 1;
            }
            if mint_data.ciphertext_validity_proof_instruction_offset != 0 {
                let account_index = *account_indexes.get(offset).ok_or(
                    ParseInstructionError::InstructionKeyMismatch(ParsableProgram::SplToken),
                )?;
                map.insert(
                    "ciphertextValidityProofRecordAccount".to_string(),
                    json!(account_keys[account_index as usize].to_string()),
                );
                offset += 1;
            }
            if mint_data.range_proof_instruction_offset != 0 {
                let account_index = *account_indexes.get(offset).ok_or(
                    ParseInstructionError::InstructionKeyMismatch(ParsableProgram::SplToken),
                )?;
                map.insert(
                    "rangeProofRecordAccount".to_string(),
                    json!(account_keys[account_index as usize].to_string()),
                );
                offset += 1;
            }
```
I might be missing something, but what do you think about this approach? If the proof offset is 0, then it is guaranteed that a record account is provided for that proof type, so we add it. If the offset is not 0, then a zk proof instruction must be included in the same transaction, but we don't parse that instruction here in the token submodule.

File: transaction-status/src/parse_token/extension/confidential_mint_burn.rs
Comment (samkim-crypto) [inline transaction-status/src/parse_token/extension/confidential_mint_burn.rs]:
```suggestion
            if mint_data.equality_proof_instruction_offset != 0 {
                let account_index = *account_indexes.get(offset).ok_or(
                    ParseInstructionError::InstructionKeyMismatch(ParsableProgram::SplToken),
                )?;
                map.insert(
                    "equalityProofRecordAccount".to_string(),
                    json!(account_keys[account_index as usize].to_string()),
                );
                offset += 1;
            }
            if mint_data.ciphertext_validity_proof_instruction_offset != 0 {
                let account_index = *account_indexes.get(offset).ok_or(
                    ParseInstructionError::InstructionKeyMismatch(ParsableProgram::SplToken),
                )?;
                map.insert(
                    "ciphertextValidityProofRecordAccount".to_string(),
                    json!(account_keys[account_index as usize].to_string()),
                );
                offset += 1;
            }
            if mint_data.range_proof_instruction_offset != 0 {
                let account_index = *account_indexes.get(offset).ok_or(
                    ParseInstructionError::InstructionKeyMismatch(ParsableProgram::SplToken),
                )?;
                map.insert(
                    "rangeProofRecordAccount".to_string(),
                    json!(account_keys[account_index as usize].to_string()),
                );
                offset += 1;
            }
```
If I am not missing something from above. Likewise, if I am not missing something, let's try to do this?

**Review by samkim-crypto — COMMENTED** [review]
Everything minus the `transaction-status` part is straightforward and looks good!

Regarding `transaction-status`, I commented on the confidential mint-burn extension. Let me know what you think because I could be missing something. If agreed, then the same comments should also apply for the confidential transfer parts with the proof offsets.

The regular confidential transfer instruction parsing logic is also a bit outdated. For example, `UpdateMint` checks for the wrong number of accounts and `ConfigureAccount` does not take into account pubkey validity proof.

I think other instructions that require zkp (withdraw, empty account, withdraw withheld tokens) also should be updated since the current parsing logic does not take into account context state accounts.

Sorry this is my fault for not keeping this part of the code base up-to-date. I would fix the rest on a follow-up PR, but since this is to be backported, we should probably fix it in this PR. Since this is a bit tedious, so I can help out and make commits to this PR if it is okay. Then we can get someone to help approve the PR. What do you think?

File: transaction-status/src/parse_token/extension/confidential_mint_burn.rs
Comment (joncinque) [reply transaction-status/src/parse_token/extension/confidential_mint_burn.rs:59]: Maybe I misunderstood the code, but it looks like this just has the mint, instructions sysvar or context state account, and authority, which means that this check is all we need. Is that not the case?

File: transaction-status/src/parse_token/extension/confidential_mint_burn.rs
Comment (joncinque) [reply transaction-status/src/parse_token/extension/confidential_mint_burn.rs]: That makes sense to me, done!

File: transaction-status/src/parse_token/extension/confidential_mint_burn.rs
Comment (joncinque) [reply transaction-status/src/parse_token/extension/confidential_mint_burn.rs]: This makes sense to me, except I think the code is backwards. We should add the account if the offset *is* zero as you wrote, right?

File: transaction-status/src/parse_token/extension/confidential_mint_burn.rs
Comment (joncinque) [reply transaction-status/src/parse_token/extension/confidential_mint_burn.rs]: Yeah this works! I just think it's backwards again here.

**Review by joncinque — COMMENTED** [review]
Thanks for the feedback! I think I got all the instructions you mentioned, but I wasn't sure what to do for the record account for a few of them. Those are marked with a TODO -- let me know what you think!

File: transaction-status/src/parse_token/extension/confidential_mint_burn.rs
Comment (samkim-crypto) [reply transaction-status/src/parse_token/extension/confidential_mint_burn.rs:59]: Oh sorry! I meant to write this comment on the `Mint` instruction, not for `RotateSupplyPubkey`. For instructions that contain multiple proof instructions, we have variable length number of accounts, so we cannot guarantee safe array indexing just with `check_num_token_accounts`. It seems like you already fixed this below!

**Comment by joncinque** [comment]
Ok, this should be ready for another look based on our most recent discussion

File: transaction-status/src/parse_token/extension/confidential_mint_burn.rs
Comment (samkim-crypto) [inline transaction-status/src/parse_token/extension/confidential_mint_burn.rs:122]: Should we add a short clarifying comment for the future saying that we assume that proof accounts are always used?

File: transaction-status/src/parse_token/extension/confidential_mint_burn.rs
Comment (samkim-crypto) [inline transaction-status/src/parse_token/extension/confidential_mint_burn.rs]: Here, technically, the proof account can either be context state account or record account, so we should either just use `equalityProofAccount` or make the distinction as follows:
```suggestion
                let label = if mint_data.equality_proof_instruction_offset == 0 {
                    "equalityProofContextStateAccount"
                } else {
                    "equalityProofRecordAccount"
                };
                map.insert(
                    label.to_string(),
                    json!(account_keys[account_indexes[offset] as usize].to_string()),
                );
```
It is your call as to whether we want to keep things simple or precise.

File: transaction-status/src/parse_token/extension/confidential_transfer.rs
Comment (samkim-crypto) [inline transaction-status/src/parse_token/extension/confidential_transfer.rs:38]: I can't really comment on non-changed part of the code, but for `InitializeMint` instruction data, we should remove `mint` and add `confidentialTransferMintAuthority,`, `autoApproveNewAccounts`, and `auditorElGamalPubkey`. Sorry I missed this on my previous review.

File: transaction-status/src/parse_token/extension/confidential_transfer.rs
Comment (samkim-crypto) [inline transaction-status/src/parse_token/extension/confidential_transfer.rs:156]: The function `parse_signers` uses unsafe indexing, but `check_num_token_accounts` does not guarantee safety if `offset` goes up to 3. Should we add an if statement to prevent panic? This seem to apply for all instructions with variable number of accounts...

File: transaction-status/src/parse_token/extension/confidential_transfer_fee.rs
Comment (samkim-crypto) [inline transaction-status/src/parse_token/extension/confidential_transfer_fee.rs:44]: I can't comment on unchanged changes, but the `InitializeConfidentialTransferFeeConfig` might be missing the instruction data in json?

**Review by samkim-crypto — COMMENTED** [review]
Looks good! Just some things that I missed from the last review and some comments. Let me know what you think!

File: transaction-status/src/parse_token/extension/confidential_mint_burn.rs
Comment (joncinque) [reply transaction-status/src/parse_token/extension/confidential_mint_burn.rs:122]: Yeah, that makes sense, will do!

File: transaction-status/src/parse_token/extension/confidential_mint_burn.rs
Comment (joncinque) [reply transaction-status/src/parse_token/extension/confidential_mint_burn.rs]: It's a good point -- we may as well extract as much information as possible. I'll update them all

File: transaction-status/src/parse_token/extension/confidential_transfer.rs
Comment (joncinque) [reply transaction-status/src/parse_token/extension/confidential_transfer.rs:38]: Will do

File: transaction-status/src/parse_token/extension/confidential_transfer_fee.rs
Comment (joncinque) [reply transaction-status/src/parse_token/extension/confidential_transfer_fee.rs:44]: Yep, done!

File: transaction-status/src/parse_token/extension/confidential_transfer.rs
Comment (joncinque) [reply transaction-status/src/parse_token/extension/confidential_transfer.rs:156]: We should be safe since we check that there's at least one more account everywhere. Here, for example, the offset `3` is only returned if there's at least 4 accounts, so it's safe. And for the offset `2`, we've already checked that there's at least `3` accounts

Comment (joncinque) [reply transaction-status/src/parse_token/extension/confidential_transfer.rs:156]: Is there a place where that isn't done?

File: transaction-status/src/parse_token/extension/confidential_transfer_fee.rs
Comment (joncinque) [reply transaction-status/src/parse_token/extension/confidential_transfer_fee.rs:44]: Done in 0d50498a99

File: transaction-status/src/parse_token/extension/confidential_transfer.rs
Comment (joncinque) [reply transaction-status/src/parse_token/extension/confidential_transfer.rs:38]: Done in 0d50498a99

File: transaction-status/src/parse_token/extension/confidential_mint_burn.rs
Comment (joncinque) [reply transaction-status/src/parse_token/extension/confidential_mint_burn.rs]: Done in 234119d974

File: transaction-status/src/parse_token/extension/confidential_mint_burn.rs
Comment (joncinque) [reply transaction-status/src/parse_token/extension/confidential_mint_burn.rs:122]: Done in 1f279e0b6e

File: transaction-status/src/parse_token/extension/confidential_transfer.rs
Comment (samkim-crypto) [reply transaction-status/src/parse_token/extension/confidential_transfer.rs:156]: Oh yes you are right! It seems that in all the places, it is safe.

**Review by samkim-crypto — DISMISSED** [review]
Looks good to me!

**Comment by joncinque** [comment]
Can I get another approval? I had to rebase

**Comment by anza-team** [comment]
:scream: New commits were pushed while the automerge label was present.

**Review by samkim-crypto — APPROVED** [review]

**Pull request merged**
